File: src/main/java/at/bitfire/ical4android/Task.java
/*
 * Copyright © 2013 – 2015 Ricki Hirner (bitfire web engineering).
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the GNU Public License v3.0
 * which accompanies this distribution, and is available at
 * http://www.gnu.org/licenses/gpl.html
 */

package at.bitfire.ical4android;

import android.util.Log;

import net.fortuna.ical4j.data.CalendarOutputter;
import net.fortuna.ical4j.data.ParserException;
import net.fortuna.ical4j.model.Calendar;
import net.fortuna.ical4j.model.Component;
import net.fortuna.ical4j.model.ComponentList;
import net.fortuna.ical4j.model.DateTime;
import net.fortuna.ical4j.model.Property;
import net.fortuna.ical4j.model.PropertyList;
import net.fortuna.ical4j.model.TimeZone;
import net.fortuna.ical4j.model.ValidationException;
import net.fortuna.ical4j.model.component.VToDo;
import net.fortuna.ical4j.model.property.Clazz;
import net.fortuna.ical4j.model.property.Completed;
import net.fortuna.ical4j.model.property.Created;
import net.fortuna.ical4j.model.property.Description;
import net.fortuna.ical4j.model.property.DtStart;
import net.fortuna.ical4j.model.property.Due;
import net.fortuna.ical4j.model.property.Duration;
import net.fortuna.ical4j.model.property.ExDate;
import net.fortuna.ical4j.model.property.Geo;
import net.fortuna.ical4j.model.property.LastModified;
import net.fortuna.ical4j.model.property.Location;
import net.fortuna.ical4j.model.property.Organizer;
import net.fortuna.ical4j.model.property.PercentComplete;
import net.fortuna.ical4j.model.property.Priority;
import net.fortuna.ical4j.model.property.RDate;
import net.fortuna.ical4j.model.property.RRule;
import net.fortuna.ical4j.model.property.Sequence;
import net.fortuna.ical4j.model.property.Status;
import net.fortuna.ical4j.model.property.Summary;
import net.fortuna.ical4j.model.property.Uid;
import net.fortuna.ical4j.model.property.Url;
import net.fortuna.ical4j.model.property.Version;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.net.URI;
import java.net.URISyntaxException;
import java.nio.charset.Charset;
import java.util.HashSet;
import java.util.LinkedList;
import java.util.List;
import java.util.Set;

import lombok.Cleanup;
import lombok.Getter;
import lombok.NonNull;

public class Task extends iCalendar {
	private final static String TAG = "ical4android.Task";

	public Long createdAt, lastModified;

    public String summary, location, description, url;
    public Organizer organizer;
    public Geo geoPosition;
    public int priority;
    public Clazz classification;
    public Status status;

    public DtStart dtStart;
    public Due due;
    public Duration duration;
    public Completed completedAt;
    public Integer percentComplete;

    public RRule rRule;
    @Getter private List<RDate> rDates = new LinkedList<>();
    @Getter private List<ExDate> exDates = new LinkedList<>();


    /**
     * Parses an InputStream that contains iCalendar VTODOs.
     *
     * @param stream  input stream containing the VTODOs
     * @param charset charset of the input stream or null (will assume UTF-8)
     * @return array of filled Task data objects (may have size 0) – doesn't return null
     * @throws IOException
     * @throws InvalidCalendarException on parser exceptions
     */
	public static Task[] fromStream(@NonNull InputStream stream, Charset charset) throws IOException, InvalidCalendarException {
        final Calendar ical;
        try {
            if (charset != null) {
                @Cleanup InputStreamReader reader = new InputStreamReader(stream, charset);
                ical = calendarBuilder.build(reader);
            } else
                ical = calendarBuilder.build(stream);
        } catch (ParserException e) {
            throw new InvalidCalendarException("Couldn't parse calendar resource", e);
        }

        ComponentList todos = ical.getComponents(Component.VTODO);
        Task[] tasks = new Task[todos.size()];
        for (int i = todos.size() - 1; i >= 0; i--)
            tasks[i] = fromVToDo((VToDo)todos.get(i));
        return tasks;
    }


    protected static Task fromVToDo(VToDo todo) throws InvalidCalendarException {
        final Task t = new Task();

		if (todo.getUid() != null)
			t.uid = todo.getUid().getValue();
		else {
			Log.w(TAG, "Received VTODO without UID, generating new one");
			t.generateUID();
		}
        if (todo.getSequence() != null)
            t.sequence = todo.getSequence().getSequenceNo();

		if (todo.getCreated() != null)
			t.createdAt = todo.getCreated().getDateTime().getTime();
		if (todo.getLastModified() != null)
			t.lastModified = todo.getLastModified().getDateTime().getTime();

		if (todo.getSummary() != null)
			t.summary = todo.getSummary().getValue();
		if (todo.getLocation() != null)
			t.location = todo.getLocation().getValue();
        if (todo.getGeographicPos() != null)
            t.geoPosition = todo.getGeographicPos();
		if (todo.getDescription() != null)
            t.description = todo.getDescription().getValue();
		if (todo.getUrl() != null)
            t.url = todo.getUrl().getValue();
        if (todo.getOrganizer() != null)
            t.organizer = todo.getOrganizer();

        t.priority = (todo.getPriority() != null) ? todo.getPriority().getLevel() : 0;
		if (todo.getClassification() != null)
            t.classification = todo.getClassification();
		if (todo.getStatus() != null)
            t.status = todo.getStatus();

		if (todo.getDue() != null) {
            t.due = todo.getDue();
			validateTimeZone(t.due);
		}
		if (todo.getDuration() != null)
            t.duration = todo.getDuration();
		if (todo.getStartDate() != null) {
            t.dtStart = todo.getStartDate();
			validateTimeZone(t.dtStart);
		}
		if (todo.getDateCompleted() != null) {
            t.completedAt = todo.getDateCompleted();
			validateTimeZone(t.completedAt);
		}
		if (todo.getPercentComplete() != null)
            t.percentComplete = todo.getPercentComplete().getPercentage();

        t.rRule = (RRule)todo.getProperty(Property.RRULE);
        for (RDate rdate : (List<RDate>) (List<?>) todo.getProperties(Property.RDATE))
            t.rDates.add(rdate);
        for (ExDate exdate : (List<ExDate>) (List<?>) todo.getProperties(Property.EXDATE))
            t.exDates.add(exdate);

        return t;
	}


	public ByteArrayOutputStream toStream() throws IOException {
		final net.fortuna.ical4j.model.Calendar ical = new net.fortuna.ical4j.model.Calendar();
		ical.getProperties().add(Version.VERSION_2_0);
		ical.getProperties().add(prodId);

		final VToDo todo = new VToDo();
		ical.getComponents().add(todo);
		final PropertyList props = todo.getProperties();

		if (uid != null)
			props.add(new Uid(uid));
        if (sequence != 0)
            props.add(new Sequence(sequence));

		if (createdAt != null)
			props.add(new Created(new DateTime(createdAt)));
		if (lastModified != null)
			props.add(new LastModified(new DateTime(lastModified)));

		if (summary != null)
			props.add(new Summary(summary));
		if (location != null)
			props.add(new Location(location));
        if (geoPosition != null)
            props.add(geoPosition);
		if (description != null)
			props.add(new Description(description));
		if (url != null)
			try {
				props.add(new Url(new URI(url)));
			} catch (URISyntaxException e) {
				Log.e(TAG, "Ignoring invalid task URL: " + url, e);
			}
        if (organizer != null)
            props.add(organizer);

		if (priority != 0)
			props.add(new Priority(priority));
		if (classification != null)
			props.add(classification);
		if (status != null)
			props.add(status);

        if (rRule != null)
            props.add(rRule);
        for (RDate rDate : rDates)
            props.add(rDate);
        for (ExDate exDate : exDates)
            props.add(exDate);

		// remember used time zones
		Set<TimeZone> usedTimeZones = new HashSet<>();
		if (due != null) {
			props.add(due);
			if (due.getTimeZone() != null)
				usedTimeZones.add(due.getTimeZone());
		}
		if (duration != null)
			props.add(duration);
		if (dtStart != null) {
			props.add(dtStart);
			if (dtStart.getTimeZone() != null)
				usedTimeZones.add(dtStart.getTimeZone());
		}
		if (completedAt != null) {
			props.add(completedAt);
			if (completedAt.getTimeZone() != null)
				usedTimeZones.add(completedAt.getTimeZone());
		}
		if (percentComplete != null)
			props.add(new PercentComplete(percentComplete));

		// add VTIMEZONE components
		for (TimeZone timeZone : usedTimeZones)
			ical.getComponents().add(timeZone.getVTimeZone());

		CalendarOutputter output = new CalendarOutputter(false);
		ByteArrayOutputStream os = new ByteArrayOutputStream();
		try {
			output.output(ical, os);
		} catch (ValidationException e) {
			Log.e(TAG, "Couldn't generate valid VTODO", e);
		}
		return os;
	}


    public boolean isAllDay() {
        return  (dtStart != null && !(dtStart.getDate() instanceof DateTime)) ||
                (due != null && !(due.getDate() instanceof DateTime));
    }

    public java.util.TimeZone getTimeZone() {
        java.util.TimeZone tz = null;
        if (dtStart != null && dtStart.getTimeZone() != null)
            tz = dtStart.getTimeZone();
        if (tz == null && due != null)
            tz = due.getTimeZone();

        // fallback
        if (tz == null)
            tz = TimeZone.getDefault();

        return tz;
    }

}

File: src/main/java/at/bitfire/ical4android/Task.kt
/***************************************************************************************************
 * Copyright © All Contributors. See LICENSE and AUTHORS in the root directory for details.
 **************************************************************************************************/

package at.bitfire.ical4android

import androidx.annotation.IntRange
import at.bitfire.ical4android.util.DateUtils
import net.fortuna.ical4j.data.CalendarOutputter
import net.fortuna.ical4j.data.ParserException
import net.fortuna.ical4j.model.*
import net.fortuna.ical4j.model.Calendar
import net.fortuna.ical4j.model.TimeZone
import net.fortuna.ical4j.model.component.VAlarm
import net.fortuna.ical4j.model.component.VToDo
import net.fortuna.ical4j.model.property.*
import java.io.IOException
import java.io.OutputStream
import java.io.Reader
import java.net.URI
import java.net.URISyntaxException
import java.util.*
import java.util.logging.Level

class Task: ICalendar() {

    var createdAt: Long? = null
    var lastModified: Long? = null

    var summary: String? = null
    var location: String? = null
    var geoPosition: Geo? = null
    var description: String? = null
    var color: Int? = null
    var url: String? = null
    var organizer: Organizer? = null

    @IntRange(from = 0, to = 9)
    var priority: Int = Priority.UNDEFINED.level

    var classification: Clazz? = null
    var status: Status? = null

    var dtStart: DtStart? = null
    var due: Due? = null
    var duration: Duration? = null
    var completedAt: Completed? = null

    @IntRange(from = 0, to = 100)
    var percentComplete: Int? = null

    var rRule: RRule? = null
    val rDates = LinkedList<RDate>()
    val exDates = LinkedList<ExDate>()

    val categories = LinkedList<String>()
    var relatedTo = LinkedList<RelatedTo>()
    val unknownProperties = LinkedList<Property>()

    val alarms = LinkedList<VAlarm>()

    companion object {

        /**
         * Parses an iCalendar resource, applies [ICalPreprocessor] to increase compatibility
         * and extracts the VTODOs.
         *
         * @param reader where the iCalendar is taken from
         *
         * @return array of filled [Task] data objects (may have size 0)
         *
         * @throws ParserException when the iCalendar can't be parsed
         * @throws IllegalArgumentException when the iCalendar resource contains an invalid value
         * @throws IOException on I/O errors
         */
        @UsesThreadContextClassLoader
        fun tasksFromReader(reader: Reader): List<Task> {
            val ical = fromReader(reader)
            val vToDos = ical.getComponents<VToDo>(Component.VTODO)
            return vToDos.mapTo(LinkedList()) { this.fromVToDo(it) }
        }

        private fun fromVToDo(todo: VToDo): Task {
            val t = Task()

            if (todo.uid != null)
                t.uid = todo.uid.value
            else {
                Ical4Android.log.warning("Received VTODO without UID, generating new one")
                t.generateUID()
            }

            // sequence must only be null for locally created, not-yet-synchronized events
            t.sequence = 0

            for (prop in todo.properties)
                when (prop) {
                    is Sequence -> t.sequence = prop.sequenceNo
                    is Created -> t.createdAt = prop.dateTime.time
                    is LastModified -> t.lastModified = prop.dateTime.time
                    is Summary -> t.summary = prop.value
                    is Location -> t.location = prop.value
                    is Geo -> t.geoPosition = prop
                    is Description -> t.description = prop.value
                    is Color -> t.color = Css3Color.fromString(prop.value)?.argb
                    is Url -> t.url = prop.value
                    is Organizer -> t.organizer = prop
                    is Priority -> t.priority = prop.level
                    is Clazz -> t.classification = prop
                    is Status -> t.status = prop
                    is Due -> { t.due = prop }
                    is Duration -> t.duration = prop
                    is DtStart -> { t.dtStart = prop }
                    is Completed -> { t.completedAt = prop }
                    is PercentComplete -> t.percentComplete = prop.percentage
                    is RRule -> t.rRule = prop
                    is RDate -> t.rDates += prop
                    is ExDate -> t.exDates += prop
                    is Categories ->
                        for (category in prop.categories)
                            t.categories += category
                    is RelatedTo -> t.relatedTo.add(prop)
                    is Uid, is ProdId, is DtStamp -> { /* don't save these as unknown properties */ }
                    else -> t.unknownProperties += prop
                }

            t.alarms.addAll(todo.alarms)

            // There seem to be many invalid tasks out there because of some defect clients, do some validation.
            val dtStart = t.dtStart
            val due = t.due

            if (dtStart != null && due != null) {
                if (DateUtils.isDate(dtStart) && DateUtils.isDateTime(due)) {
                    Ical4Android.log.warning("DTSTART is DATE but DUE is DATE-TIME, rewriting DTSTART to DATE-TIME")
                    t.dtStart = DtStart(DateTime(dtStart.value, due.timeZone))
                } else if (DateUtils.isDateTime(dtStart) && DateUtils.isDate(due)) {
                    Ical4Android.log.warning("DTSTART is DATE-TIME but DUE is DATE, rewriting DUE to DATE-TIME")
                    t.due = Due(DateTime(due.value, dtStart.timeZone))
                }


                if (due.date <= dtStart.date) {
                    Ical4Android.log.warning("Found invalid DUE <= DTSTART; dropping DTSTART")
                    t.dtStart = null
                }
            }

            if (t.duration != null && t.dtStart == null) {
                Ical4Android.log.warning("Found DURATION without DTSTART; ignoring")
                t.duration = null
            }

            return t
        }

    }


    @UsesThreadContextClassLoader
    fun write(os: OutputStream) {
        Ical4Android.checkThreadContextClassLoader()

        val ical = Calendar()
        ical.properties += Version.VERSION_2_0
        ical.properties += prodId

        val vTodo = VToDo(true /* generates DTSTAMP */)
        ical.components += vTodo
        val props = vTodo.properties

        uid?.let { props += Uid(uid) }
        sequence?.let {
            if (it != 0)
                props += Sequence(it)
        }

        createdAt?.let { props += Created(DateTime(it)) }
        lastModified?.let { props += LastModified(DateTime(it)) }

        summary?.let { props += Summary(it) }
        location?.let { props += Location(it) }
        geoPosition?.let { props += it }
        description?.let { props += Description(it) }
        color?.let { props += Color(null, Css3Color.nearestMatch(it).name) }
        url?.let {
            try {
                props += Url(URI(it))
            } catch (e: URISyntaxException) {
                Ical4Android.log.log(Level.WARNING, "Ignoring invalid task URL: $url", e)
            }
        }
        organizer?.let { props += it }

        if (priority != Priority.UNDEFINED.level)
            props += Priority(priority)
        classification?.let { props += it }
        status?.let { props += it }

        rRule?.let { props += it }
        rDates.forEach { props += it }
        exDates.forEach { props += it }

        if (categories.isNotEmpty())
            props += Categories(TextList(categories.toTypedArray()))
        props.addAll(relatedTo)
        props.addAll(unknownProperties)

        // remember used time zones
        val usedTimeZones = HashSet<TimeZone>()
        due?.let {
            props += it
            it.timeZone?.let(usedTimeZones::add)
        }
        duration?.let(props::add)
        dtStart?.let {
            props += it
            it.timeZone?.let(usedTimeZones::add)
        }
        completedAt?.let {
            props += it
            it.timeZone?.let(usedTimeZones::add)
        }
        percentComplete?.let { props += PercentComplete(it) }

        if (alarms.isNotEmpty())
            vTodo.alarms.addAll(alarms)

        // determine earliest referenced date
        val earliest = arrayOf(
                dtStart?.date,
                due?.date,
                completedAt?.date
        ).filterNotNull().minOrNull()
        // add VTIMEZONE components
        for (tz in usedTimeZones)
            ical.components += minifyVTimeZone(tz.vTimeZone, earliest)

        softValidate(ical)
        CalendarOutputter(false).output(ical, os)
    }


    fun isAllDay(): Boolean {
        return  dtStart?.let { DateUtils.isDate(it) } ?:
                due?.let { DateUtils.isDate(it) } ?:
                true
    }

}
